### PR TITLE
fix: MLSConferenceStaleParticipantsRemover not removed after end of call - WPB-5340

### PR DESCRIPTION
**Pull request by netbe**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-5340" title="WPB-5340" target="_blank">BugWPB-5340</a>  [iOS] MLS: Not possible to enable video during MLS Federated group call
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Working on calling video issues, we're getting some errors while getting subconversation client ids from CoreCrypto (which is used to update the participants state).

### Causes (Optional)

The MLSConferenceStaleParticipantsRemover is a custom subscriber that demands events forever to the publisher.

### Solutions

Stop the subscription of MLSConferenceStaleParticipantsRemover when the call ends.

### Notes (Optional)

Also fixes a CoreData threading violation

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift [hunk 1]
```diff
@@ -48,9 +48,29 @@ class MLSConferenceStaleParticipantsRemover: Subscriber {
...
+                subscription?.cancel()
+                subscription = nil
+            }
+        }
+    }
+
     func receive(subscription: Subscription) {
+        self.subscription = subscription
         subscription.request(.unlimited)
     }
```
Comment (David-Henner) [inline wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift]: Is there a point to using `didset` here?
Maybe I'm missing something, but wouldn't this be simpler?


```suggestion
    deinit {
        stopSubscribing()
    }

    private var subscription: Subscription? 
    
    // MARK: - Subscriber implementation
    
    func stopSubscribing() {
        subscription?.cancel()
        subscription = nil
    }

    func receive(subscription: Subscription) {
        self.subscription = subscription
        subscription.request(.unlimited)
    }
```

**Review by markusfassbender — APPROVED** [review]
Approve, just the linter complaints..

File: wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift [hunk 2]
```diff
@@ -48,9 +48,21 @@ class MLSConferenceStaleParticipantsRemover: Subscriber {
         self.removalTimeout = removalTimeout
     }
 
+    deinit {
+        stopSubscribing()
+    }
+
```
Comment (David-Henner) [inline wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift]:
```suggestion

```

File: wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift [hunk 3]
```diff
@@ -48,9 +48,21 @@ class MLSConferenceStaleParticipantsRemover: Subscriber {
         self.removalTimeout = removalTimeout
     }
 
+    deinit {
+        stopSubscribing()
+    }
+    
+    private var subscription: Subscription?
```
Comment (David-Henner) [inline wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift]:
```suggestion
    private var subscription: Subscription?
```

File: wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift [hunk 4]
```diff
@@ -48,9 +48,21 @@ class MLSConferenceStaleParticipantsRemover: Subscriber {
         self.removalTimeout = removalTimeout
     }
 
+    deinit {
+        stopSubscribing()
+    }
+    
+    private var subscription: Subscription? 
+
```
Comment (David-Henner) [inline wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift]:
```suggestion

```

File: wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift [hunk 5]
```diff
@@ -48,9 +48,21 @@ class MLSConferenceStaleParticipantsRemover: Subscriber {
...
     }
 
+    deinit {
+        stopSubscribing()
+    }
+    
+    private var subscription: Subscription? 
+    
     // MARK: - Subscriber implementation
+
```
Comment (David-Henner) [inline wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift]:
```suggestion

```

**Review by David-Henner — APPROVED** [review]
Approved.
You can make a batch commit from my comments, it should fix linting issues

**Pull request merged**